### ISSUE TITLE
OpenSSL link error on brightbox Ruby 2.5.5 and Ubuntu Trusty

### ISSUE THREAD

**Issue by julik** [puma/puma#1833]
### Steps to reproduce

1) On a Trusty 14.04 install, install Ruby 2.5.5 and puma 4.0 and try to start it

### Expected behavior

Puma starts

### Actual behavior

Puma crashes on startup with an undefined symbol error:

```
2019-07-02T10:27:57.293554+00:00 backend-s-eu-0737cb4504ece8209.node.eu-west-1.wt kernel: [4056975.003502] init: puma main process ended, respawning
2019-07-02T10:27:59.597406+00:00 backend-s-eu-0737cb4504ece8209.node.eu-west-1.wt puma: `/` is not writable.
2019-07-02T10:27:59.597420+00:00 backend-s-eu-0737cb4504ece8209.node.eu-west-1.wt puma: Bundler will use `/tmp/bundler/home/unknown' as your home directory temporarily.
2019-07-02T10:27:59.800240+00:00 backend-s-eu-0737cb4504ece8209.node.eu-west-1.wt puma: <app_path>/bundle/ruby/2.5.0/gems/puma-4.0.0/lib/puma/server.rb:16:in `require': <app_path>/bundle/ruby/2.5.0/gems/puma-4.0.0/lib/puma/puma_http11.so: undefined symbol: DTLS_method - <app_path>/bundle/ruby/2.5.0/gems/puma-4.0.0/lib/puma/puma_http11.so (LoadError)
2019-07-02T10:27:59.800299+00:00 backend-s-eu-0737cb4504ece8209.node.eu-west-1.wt puma: #011from <app_path>/bundle/ruby/2.5.0/gems/puma-4.0.0/lib/puma/server.rb:16:in `<top (required)>'
2019-07-02T10:27:59.800317+00:00 backend-s-eu-0737cb4504ece8209.node.eu-west-1.wt puma: #011from <app_path>/bundle/ruby/2.5.0/gems/puma-4.0.0/lib/puma/runner.rb:3:in `require'
2019-07-02T10:27:59.800331+00:00 backend-s-eu-0737cb4504ece8209.node.eu-west-1.wt puma: #011from <app_path>/bundle/ruby/2.5.0/gems/puma-4.0.0/lib/puma/runner.rb:3:in `<top (required)>'
2019-07-02T10:27:59.800345+00:00 backend-s-eu-0737cb4504ece8209.node.eu-west-1.wt puma: #011from
```

### System configuration

**Ruby version**: ruby 2.5.5p157 (2019-03-15 revision 67260) [x86_64-linux-gnu]
**Rails version**: None used
**Puma version**: 4.0.0

**Comment by MSP-Greg** [comment]
See #1832, which fixes a older PR by me...

At present, all Puma CI testing is using OpenSSL 1.0.2 or later.  Trusty uses 1.0.1, which, JFYI, is no longer supported.

`DTLS_method` is not available in some versions of OpenSSL 1.0.1.  I did run a test on Ruby 2.0.x on Trusty, and it passed with the PR...

**Comment by MSP-Greg** [comment]
@nateberkopec, et al,

This does beg the question of whether Puma should build/run with older OpenSSL versions, should it fail, post a warning, etc...

**Comment by MSP-Greg** [comment]
@nateberkopec 

Looked in the wrong place earlier...   With a scratch branch, I ran Trusty 2.2 on Travis, passed, job is here:
https://travis-ci.org/MSP-Greg/puma/jobs/553015028

Want a commit with one Trusty/1.0.1 job added to Travis?

**Comment by julik** [comment]
Great that it could be found so quickly

**Comment by MSP-Greg** [comment]
@nateberkopec 

Locally, I added a commit to the branch for #1832 
```
Add extconf test for TLS_server_method & use in mini_ssl.c
```

When Travis gets back to normal, I'll push it.

With that change, most of the changes left in #1788 & #1799 are new constants like `OPENSSL_NO_SSL3` and `OPENSSL_NO_TLS1`, along with implementing

```
SSL_CTX_set_min_proto_version
```

which is new in OpenSSL 1.1.0.  And also tests, which I'd like to review, and consider maybe adding more.

Any thoughts?  I'm a little busy right now, so I'd like to wait on creating the new PR from #1788 & #1799...

**Comment by nateberkopec** [comment]
Fixed by #1832
